example/echo-op/echo_op.cpp

```diff
@@ -14,276 +14,324 @@
-    //
-    static_assert(boost::beast::is_async_stream<AsyncStream>::value,
-        "AsyncStream requirements not met");
+        void
+        operator()(
+            beast::error_code ec,
+            std::size_t bytes_transferred = 0,
+            bool cont = true) /*< Second and subsequent invocations will seee `cont=true`. */
+        {
+            // The `reenter` keyword transfers control to the last
+            // yield point, or to the beginning of the scope if
+            // this is the first time.
+
+            reenter(*this)
+            {
+                for(;;)
+                {
+                    std::size_t pos;
+
+                    // Search for a newline in the readable bytes of the buffer
+                    pos = find_newline(buffer_.data());
+
+                    // If we don't have the newline, then read more
+                    if(pos == 0)
+                    {
+                        std::size_t bytes_to_read;
+
+                        // Determine the number of bytes to read,
+                        // using available capacity in the buffer first.
+
+                        bytes_to_read = std::min<std::size_t>(
+                              std::max<std::size_t>(512,                // under 512 is too little,
+                                  buffer_.capacity() - buffer_.size()),
+                              std::min<std::size_t>(65536,              // and over 65536 is too much.
+                                  buffer_.max_size() - buffer_.size()));
+
+                        // Read some data into our dynamic buffer_. We transfer
+                        // ownership of the composed operation by using the
+                        // `std::move(*this)` idiom. The `yield` keyword causes
+                        // the function to return immediately after the initiating
+                        // function returns.
+
+                        yield stream_.async_read_some(
+                            buffer_.prepare(bytes_to_read), std::move(*this));
+
+                        // After the `async_read_some` completes, control is
+                        // transferred to this line by the `reenter` keyword.
+
+                        // Move the bytes read from the writable area to the
+                        // readable area.
+
+                        buffer_.commit(bytes_transferred);
+
+                        // If an error occurs, deliver it to the caller's completion handler.
+                        if(ec)
+                            break;
+
+                        // Keep looping until we get the newline
+                        continue;
+                    }
+
+                    // We have our newline, so send the first `pos` bytes of the
+                    // buffers. The function `buffers_prefix` returns the front part
+                    // of the buffers we want.
+
+                    yield net::async_write(stream_,
```

Comment:
Can't this use `net::async_write(stream_, buffer.data(), net::transfer_exactly(pos), std::move(*this)`?

Reply:
In theory yes, but its nice to see as many beast functions as possible in one example to show their use